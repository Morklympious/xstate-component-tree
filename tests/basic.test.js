"use strict";

const { Machine : createMachine, interpret } = require("xstate");
const trees = require("./util/trees.js");
const component = require("./util/component.js");

describe("xstate-component-tree", () => {
    it("should return a tree of components", async () => {
        const testMachine = createMachine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        component : component("one"),
                    },

                    initial : "two",

                    states : {
                        two : {
                            meta : {
                                component : component("two"),
                            },
                        },
                    },
                },
            },
        });

        const service = interpret(testMachine);

        const tree = trees(service);

        expect(await tree()).toMatchSnapshot();
    });

    it("should support parallel states", async () => {
        const testMachine = createMachine({
            type : "parallel",

            states : {
                one : {
                    meta : {
                        component : component("one"),
                    },
                },

                two : {
                    meta : {
                        component : component("two"),
                    },
                },
            },
        });

        const service = interpret(testMachine);

        const tree = trees(service);

        expect(await tree()).toMatchSnapshot();
    });

    it("should support nested parallel states", async () => {
        const testMachine = createMachine({
            initial : "one",

            states : {
                one : {
                    type : "parallel",

                    states : {
                        two : {
                            meta : {
                                component : component("two"),
                            },
                        },
        
                        three : {
                            meta : {
                                component : component("three"),
                            },
                        },
                    },
                },
            },
        });

        const service = interpret(testMachine);

        const tree = trees(service);

        expect(await tree()).toMatchSnapshot();
    });

    it("should support arbitrary ids", async () => {
        const testMachine = createMachine({
            initial : "one",

            states : {
                one : {
                    id : "foo",

                    meta : {
                        component : component("one"),
                    },
                    
                    initial : "two",
                    
                    states : {
                        two : {
                            id : "bar",

                            meta : {
                                component : component("two"),
                            },
                        },
                    },
                },
            },
        });

        const service = interpret(testMachine);

        const tree = trees(service);

        expect(await tree()).toMatchSnapshot();
    });

    it("should support holes", async () => {
        const testMachine = createMachine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        component : component("one"),
                    },

                    initial : "two",

                    states : {
                        two : {
                            initial : "three",

                            states : {
                                three : {
                                    meta : {
                                        component : component("three"),
                                    },
                                },
                            },
                        },
                    },
                },
            },
        });

        const service = interpret(testMachine);

        const tree = trees(service);

        expect(await tree()).toMatchSnapshot();
    });
    
    it("should rebuild on machine transition", async () => {
        const testMachine = createMachine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        component : component("one"),
                    },

                    on : {
                        NEXT : "two",
                    },
                },

                two : {
                    meta : {
                        component : component("two"),
                    },
                },
            },
        });

        const service = interpret(testMachine);

        const tree = trees(service);
        
        const before = await tree();
        
        service.send("NEXT");

        const after = await tree();

        expect(before).toMatchDiffSnapshot(after);
    });

    it("shouldn't rebuild on events without changes", async () => {
        const testMachine = createMachine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        component : component("one"),
                    },
                },
            },
        });

        const service = interpret(testMachine);
        const eventCounter = jest.fn();

        service.onEvent(eventCounter);

        const tree = trees(service);
        
        await tree();

        service.send("NEXT");

        // onEvent was called twice, but treeBuilder returned on tree as expected
        expect(eventCounter.mock.calls.length).toBe(2);
    });

    // o god broked
    it.skip("shouldn't trigger the callback if another change occurs before the tree is built", async () => {
        const testMachine = createMachine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        load : () => {
                            loader = new Promise((resolve) =>
                                setTimeout(() => resolve(component("one")), 0)
                            );

                            return loader;
                        },
                    },

                    on : {
                        NEXT : "two",
                    },
                },

                two : {
                    meta : {
                        component : component("two"),
                    },
                },
            },
        });

        const service = interpret(testMachine);
        const eventCounter = jest.fn();

        service.onEvent(eventCounter);

        const states = trees(service);

        const loader = states();

        service.send("NEXT");

        await loader;

        // onEvent was called twice, but treeBuilder returned on tree as expected
        expect(eventCounter.mock.calls.length).toBe(2);
    });

    it.todo("should support top-level machine ids in the built tree");
});
