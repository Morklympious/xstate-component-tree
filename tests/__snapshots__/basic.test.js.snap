// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xstate-component-tree should rebuild on machine transition 1`] = `
Snapshot Diff:
- First value
+ Second value

  Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
-         "component": "one",
+         "component": "two",
        },
      ],
      "id": "(machine)",
    },
  ]
`;

exports[`xstate-component-tree should return a tree of components 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "component": "two",
          },
        ],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
]
`;

exports[`xstate-component-tree should support arbitrary ids 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "component": "two",
          },
        ],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
]
`;

exports[`xstate-component-tree should support holes 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "component": "three",
          },
        ],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
]
`;

exports[`xstate-component-tree should support nested parallel states 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "two",
      },
      Object {
        "children": Array [],
        "component": "three",
      },
    ],
    "id": "(machine)",
  },
]
`;

exports[`xstate-component-tree should support parallel states 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
      Object {
        "children": Array [],
        "component": "two",
      },
    ],
    "id": "(machine)",
  },
]
`;

exports[`xstate-component-tree should support top-level machine ids in the built tree 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "testmachine",
  },
]
`;
